### ISSUE TITLE
indent-error-flow incorrectly identifying conditional attribution as redundant

### ISSUE THREAD

**Issue by sonalys** [mgechev/revive#564]
**Describe the bug**
When we declare variables inside if statements, and returns inside the if block, on the else statement, we can no longer use other declared variables because revive detects indent-error-flow redundancy.

**To Reproduce**
1. I updated revive `go get -u github.com/mgechev/revive`
2. I run it with the following flags & configuration file:
```go
if a, b := 1, 2; a != 1 {
  return a
} else {
 return b
}
```

**Expected behavior**
This should not be redundant, because we are only declaring those variables inside the if statement, and it should not be possible to access b without the else statement.

**Comment by chavacava** [comment]
Hi @sonalys, thanks for filling the issue.

For the example code you provided, `revive` should generate a message like:

```
if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
```
As the message says, if necessary, you'll need to move the variable declaration to its own line. Something like:
```go
a, b := 1, 2
if a != 1 {
  return a
} 
return b
```

The current implementation of the rule in `revive` is from the original one from `golint` and is based the on a simple _structural_ analysis (in the lines of "if the _then_ block ends with a structured jump (return, break, continue) and there is an _else_ branch then warn") 
There was an [discussion on this subject in the `golint` repo](https://github.com/golang/lint/issues/17) but the GO team closed the subject without modifying `golint` behavior.

One common critic to the refactoring proposed by `golint` is that now the scope of the variables is not constrained to the `if` branches. To alleviate that one could refactor to:

```go
{
  a, b := 1, 2
   if a != 1 {
    return a
  }  
  return b
}
```


Note: avoiding warnings in cases like those of your example will require to add a data flow analysis ("any variable defined in the `if` is used in the _then_ branch?")

**Comment by sonalys** [comment]
yes, improving the message would be great, because it doesn't seem to justify the behaviour it wants with the actual message
if it considered the declaration of variables inside the if scope it would be great, because sometimes it's really bothersome to already have an error declared above in the code, and you need a new variable like in the code below:
```go
err := returnsError()
if err != nil {
  return err
}

myNewVar, err = myFuncThatReturnsVarAndError()
if err != nil {
 return err
}
```
because we will need to either declare myNewVar outside the if-statement, or make the err variable to be scoped.
if I want to do something like
```go
if myNewVar, err := myFuncThatReturnsVarAndError(); err != nil {
   return err
} else {
  useMyNewVarIntoSomethingBecauseItsStillInScope(myNewVar)
}
```
it will not work

I think pre-declaring this stuff outside if-statements can make the code more difficult to read and bigger, an example of the first case
```go
var err error
err = returnsError()
if err != nil {
  return err
}

var myNewVar MyType
if myNewVar, err = myFuncThatReturnsVarAndError() {
...
}
```

when it could be

```go
if err := returnsError(); err != nil {
 return err
}

if myNewVar, err := myFuncThatReturnsVarAndError(); err != nil {
  return err
} else {
  useMyNewVarIntoSomethingBecauseItsStillInScope(myNewVar)
}
```

**Comment by chavacava** [comment]
As I've said, for the example you provided, `revive` **already generates** a message in the lines of:
```
if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
```
If in your case is not generating such a message, please provide the actual code you are linting.

There are many subjects that are a matter of personal taste, thus which is more easy to read
```go
if myNewVar, err := myFuncThatReturnsVarAndError(); err != nil {
   return err
} else {
  useMyNewVarIntoSomethingBecauseItsStillInScope(myNewVar)
}
```
or 
```go
myNewVar, err := myFuncThatReturnsVarAndError()
if err != nil {
   return err
} else {
  useMyNewVarIntoSomethingBecauseItsStillInScope(myNewVar)
}
```
It is not something  we will discuss :)
Again, you can _fix_ the scoping problem by creating a new scope by surrounding the declaration and the `if` statement with `{`  and `}`... whether this fix is _ugly_, _nice_, ... is up to you.
